File: connector/src/main/java/org/geysermc/connector/common/serializer/AsteriskSerializer.java
/*
 * Copyright (c) 2019-2020 GeyserMC. http://geysermc.org
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 * @author GeyserMC
 * @link https://github.com/GeyserMC/Geyser
 */

package org.geysermc.connector.common.serializer;

import com.fasterxml.jackson.annotation.JacksonAnnotationsInside;
import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.BeanProperty;
import com.fasterxml.jackson.databind.JsonSerializer;
import com.fasterxml.jackson.databind.SerializerProvider;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.fasterxml.jackson.databind.ser.ContextualSerializer;
import com.fasterxml.jackson.databind.ser.std.StdSerializer;

import java.io.IOException;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import java.util.Optional;

public class AsteriskSerializer extends StdSerializer<Object> implements ContextualSerializer {

    public static boolean showSensitive = false;

    @Target({ElementType.FIELD})
    @Retention(RetentionPolicy.RUNTIME)
    @JacksonAnnotationsInside
    @JsonSerialize(using = AsteriskSerializer.class)
    public @interface Asterisk {
        String value() default "***";
        boolean sensitive() default false;
    }

    String asterisk;
    boolean sensitive;

    public AsteriskSerializer() {
        super(Object.class);
    }

    public AsteriskSerializer(String asterisk, boolean sensitive) {
        super(Object.class);
        this.asterisk = asterisk;
        this.sensitive = sensitive;
    }

    @Override
    public JsonSerializer<?> createContextual(SerializerProvider serializerProvider, BeanProperty property) {
        Optional<Asterisk> anno = Optional.ofNullable(property)
                .map(prop -> prop.getAnnotation(Asterisk.class));

        return new AsteriskSerializer(anno.map(Asterisk::value).orElse(null), anno.map(Asterisk::sensitive).orElse(null));
    }

    @Override
    public void serialize(Object obj, JsonGenerator gen, SerializerProvider prov) throws IOException {
        if (sensitive && showSensitive) {
            gen.writeObject(obj);
            return;
        }

        gen.writeString(asterisk);
    }
}

File: core/src/main/java/org/geysermc/geyser/text/AsteriskSerializer.cpp
/*
 * Copyright (c) 2019-2022 GeyserMC. http://geysermc.org
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 * @author GeyserMC
 * @link https://github.com/GeyserMC/Geyser
 */

package org.geysermc.geyser.text;

#include "com.google.gson.JsonElement"
#include "com.google.gson.JsonPrimitive"
#include "com.google.gson.JsonSerializationContext"
#include "com.google.gson.JsonSerializer"
#include "org.spongepowered.configurate.objectmapping.meta.Processor"
#include "org.spongepowered.configurate.serialize.SerializationException"

#include "java.lang.annotation.ElementType"
#include "java.lang.annotation.Retention"
#include "java.lang.annotation.RetentionPolicy"
#include "java.lang.annotation.Target"
#include "java.lang.reflect.Type"
#include "java.net.InetAddress"

public class AsteriskSerializer implements JsonSerializer<std::string> {
    public static final String[] NON_SENSITIVE_ADDRESSES = {"", "0.0.0.0", "localhost", "127.0.0.1", "auto", "unknown"};

    public static bool showSensitive = false;

    override public JsonElement serialize(std::string src, Type typeOfSrc, JsonSerializationContext context) {
        if (showSensitive || !isSensitiveIp(src)) {
            return new JsonPrimitive(src);
        }

        return new JsonPrimitive("***");
    }

    private static bool isSensitiveIp(std::string ip) {
        for (std::string address : NON_SENSITIVE_ADDRESSES) {
            if (address.equalsIgnoreCase(ip)) {
                return false;
            }
        }

        try {
            InetAddress address = InetAddress.getByName(ip);
            if (address.isSiteLocalAddress() || address.isLoopbackAddress()) {
                return false;
            }
        } catch (Exception e) {

        }

        return true;
    }

    public static Processor.Factory<Asterisk, std::string> CONFIGURATE_SERIALIZER = (data, fieldType) -> (value, destination) -> {
        if (showSensitive || !isSensitiveIp(value)) {
            return;
        }
        try {
            destination.set("***");
        } catch (SerializationException e) {
            throw new RuntimeException("Unable to censor IP address", e);
        }
    };

    @Target({ElementType.FIELD, ElementType.METHOD})
    @Retention(RetentionPolicy.RUNTIME)
    public @interface Asterisk {

    }
}
